### PR TITLE
Create headings for apps, users, settings pages

### LINKED ISSUE #35626: [BITV] No heading for settings pages
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] Nextcloud Server **is** running on 64bit capable CPU, PHP and OS.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

If I open the settings page of Nextcloud there is no h1 heading found in the page's source.

### Steps to reproduce

1. Open `/index.php/settings/user` or `/index.php/settings/admin`
2. Navigate through the pages from the left sidebar

### Expected behavior

A h1 in the content area for the currently selected entry, e.g. "Personal info".

The heading should have tabindex="-1" so the focus jumps to it after selecting the page.

### Installation method

Official All-in-One appliance

### Operating system

RHEL/CentOS

### PHP engine version

PHP 8.1

### Web server

Other

### Database engine version

MySQL

### Is this bug present after an update or on a fresh install?

Fresh Nextcloud Server install

### Are you using the Nextcloud Server Encryption module?

Encryption is Enabled

### What user-backends are you using?

- [ ] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

_No response_

### List of activated Apps

```shell
N/a
```

### Nextcloud Signing status

_No response_

### Nextcloud Logs

_No response_

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by JuliaKirschenheuter**
* Resolves: https://github.com/nextcloud/server/issues/35626

## Summary
Create headings for settings, users and apps pages. No visible changes.

### After
Screenshot from 2023-02-13 15-45-16
Screenshot from 2023-02-13 15-45-45

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: apps/settings/src/router.js [hunk 1]
```diff
@@ -126,6 +126,7 @@ router.afterEach(async (to) => {
 	const metaTitle = await to.meta.title?.(to)
 	if (metaTitle) {
 		document.title = `${metaTitle} - ${baseTitle}`
+		document.getElementById('pageHeadingLevel2').textContent = `${t('settings', 'Subsection: ')} ${metaTitle}`
```
Comment (nickvergessen) [inline apps/settings/src/router.js:129]: Breaks translations

https://docs.nextcloud.com/server/latest/developer_manual/basics/front-end/l10n.html#improving-your-translations

File: core/templates/layout.user.php [hunk 2]
```diff
@@ -57,6 +57,12 @@
 		</div>
 
 		<header role="banner" id="header">
+			<h1 class="hidden-visually">
+				<?php p($l->t('Section').': ').p(!empty($_['application'])?$_['application']:''); ?>
```
Comment (nickvergessen) [inline core/templates/layout.user.php:61]: Breaks translations
https://docs.nextcloud.com/server/latest/developer_manual/basics/front-end/l10n.html#improving-your-translations

File: core/templates/layout.user.php [hunk 3]
```diff
@@ -57,6 +57,12 @@
...
+			</h1>
+			<h2 class="hidden-visually" id="pageHeadingLevel2">
+				<?php p($l->t('Subsection').': ').p(!empty($_['pageTitle'])?$_['pageTitle']:''); ?>
```
Comment (nickvergessen) [inline core/templates/layout.user.php:64]: Breaks translations
https://docs.nextcloud.com/server/latest/developer_manual/basics/front-end/l10n.html#improving-your-translations

File: core/templates/layout.user.php [hunk 4]
```diff
@@ -57,6 +57,12 @@
 		</div>
 
 		<header role="banner" id="header">
+			<h1 class="hidden-visually">
```
Comment (nickvergessen) [inline core/templates/layout.user.php:60]: Iirc only 1 h1 is allowed per page. T
So this makes all pages with an existing h1 not valid anymore

File: core/templates/layout.user.php (hunk 2 above)
Comment (nickvergessen) [reply core/templates/layout.user.php:61]: Also I'm really unsure about the Section and Subsection words being added. They seem to not add useful information (and bring troubles with translations and more)

I would try to not add this generic in the layout templates but fix it in the content area of the app which might even have a matching h1 and h2 already

**Review by nickvergessen — CHANGES_REQUESTED** [review]

File: core/templates/layout.user.php (hunk 4 above)
Comment (JuliaKirschenheuter) [reply core/templates/layout.user.php:60]: sure, thanks. wanted to take it away, but forgotten. Such heading with always same content make no sense.

**Comment by JuliaKirschenheuter** [comment]
@michaelnissenbaum could you please tell your opinion about naming of new headings. Or please suggest which one could sound better. Thank you!

File: apps/settings/src/router.js (hunk 1 above)
Comment (JuliaKirschenheuter) [reply apps/settings/src/router.js:129]: thanks

File: core/templates/layout.user.php (hunk 2 above)
Comment (JuliaKirschenheuter) [reply core/templates/layout.user.php:61]: ok, thanks

File: apps/settings/src/router.js [hunk 5]
```diff
@@ -27,6 +27,7 @@ import Router from 'vue-router'
 import { generateUrl } from '@nextcloud/router'
 import { APPS_SECTION_ENUM } from './constants/AppsConstants.js'
 import store from './store/index.js'
+import Accessibility from '../../../core/src/OCP/accessibility'
```
Comment (ChristophWurst) [inline apps/settings/src/router.js:30]: import the function directly

File: core/src/OCP/accessibility.js [hunk 6]
```diff
@@ -29,4 +29,15 @@ export default {
 	disableKeyboardShortcuts() {
 		return loadState('theming', 'shortcutsDisabled', false)
 	},
+	/**
+	 * Set a page title to h1 header
```
Comment (ChristophWurst) [inline core/src/OCP/accessibility.js:33]:
```suggestion
	 * Set the page heading
```

File: core/src/OCP/accessibility.js [hunk 7]
```diff
@@ -29,4 +29,15 @@ export default {
...
+	 *
+	 * @param {string} [pageTitle] page title from the history api
+	 */
+	setPageTitle(pageTitle) {
```
Comment (ChristophWurst) [inline core/src/OCP/accessibility.js:37]:
```suggestion
	setPageHeading(heading) {
```

because the title of a page is easily confused with document title

**Review by ChristophWurst — CHANGES_REQUESTED** [review]

**Comment by ChristophWurst** [comment]
The new `OCP.Accessibility` function should go into the developer documentation

File: apps/settings/src/router.js [hunk 8]
```diff
@@ -27,6 +27,7 @@ import Router from 'vue-router'
 import { generateUrl } from '@nextcloud/router'
 import { APPS_SECTION_ENUM } from './constants/AppsConstants.js'
 import store from './store/index.js'
+import Accessibility from '../../../core/src/OCP/accessibility.js'
```
Comment (ChristophWurst) [inline apps/settings/src/router.js:30]: Import only the function you need plase, not the full module

File: apps/settings/src/router.js (hunk 5 above)
Comment (JuliaKirschenheuter) [reply apps/settings/src/router.js]: in a case of core/src/OCP/accessibility.js i'm exporting default object with functions and not 2 independent functions. In this case it is not possible to make an destruction 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import

Comment (ChristophWurst) [reply apps/settings/src/router.js]: You can change the module to have a named export of the free function `setPageHeading` and also use that function in the default export object. This should help with tree shaking.

Comment (JuliaKirschenheuter) [reply apps/settings/src/router.js]: Am i resolved this issue?

File: core/src/OCP/accessibility.js [hunk 9]
```diff
@@ -22,11 +22,23 @@
 
 import { loadState } from '@nextcloud/initial-state'
 
+/**
+ * Set the page heading
+ *
+ * @param {string} [heading] page title from the history api
```
Comment (ChristophWurst) [inline core/src/OCP/accessibility.js:28]:
```suggestion
 * @param {string} heading page title from the history api
```

https://jsdoc.app/tags-param.html

File: core/src/OCP/accessibility.js [hunk 10]
```diff
@@ -22,11 +22,23 @@
...
+ */
```
Comment (ChristophWurst) [inline core/src/OCP/accessibility.js:29]: nit

```suggestion
 * @since 27.0.0
 */
```

although this should probably be dropped from the future backport PR

File: core/templates/layout.user.php [hunk 11]
```diff
@@ -57,6 +57,9 @@
 		</div>
 
 		<header role="banner" id="header">
+			<h1 class="hidden-visually" id="pageHeadingLevel1">
+				<?php p(!empty($_['pageTitle'])?$_['pageTitle']:''); ?>
```
Comment (ChristophWurst) [inline core/templates/layout.user.php:61]: I think the logic should be to fall back to `$theme->getName()` instead of an empty string if pageTitle is not set

**Review by ChristophWurst — APPROVED** [review]
Code looks good

**Comment by ChristophWurst** [comment]
Let's stick with fixups until the second approval and passing CI (with the exception of the fixup check)

Bildschirmfoto vom 2023-03-06 13-07-07

^ makes it very hard to follow your latest pushes

File: core/src/OCP/accessibility.js [hunk 12]
```diff
@@ -22,11 +22,24 @@
...
 import { loadState } from '@nextcloud/initial-state'
 
+/**
+ * Set the page heading
+ *
+ * @param {string} heading page title from the history api
+ * @since 27.0.0
+ */
+export function setPageHeading(heading) {
+	const headingEl = document.getElementById('pageHeadingLevel1')
```
Comment (artonge) [inline core/src/OCP/accessibility.js:32]: Let's use snake case for the id: `page-heading-level-1`

**Review by artonge — APPROVED** [review]
Small nitpick, else it looks good :)

**Review by Pytal — APPROVED** [review]

**Comment by DaphneMuller** [comment]
Hello!
Thank you for your work on this! I noticed that your PR still has the label 'documentation'. I'm not sure if you already had a look at the documentation for your PR, but at Nextcloud we strive to document changes that affect other app developers or other admins before the release takes place, and I'm pinging you with the hope to get this done. 

Are you familiar with our documentation process already? 
Changes that affect app developers should be documented here:
https://docs.nextcloud.com/server/27/developer_manual/app_publishing_maintenance/app_upgrade_guide/upgrade_to_27.html

Changes that affect administrators should be documented here:
https://docs.nextcloud.com/server/latest/admin_manual/release_notes/upgrade_to_27.html

While I understand sometimes it's handy to merge fast to get your changes in, it would be great if next time you could try to add the documentation before merging. You can find more information on that here:
https://docs.nextcloud.com/server/27/developer_manual/prologue/compatibility_app_ecosystem.html#documentation-procedures-of-changes-that-affect-app-developers

If all your documentation efforts are done, please remove the label 'documentation' and check the checkbox in your opening note and I'll stop bugging you :) 

Many thanks in advance and thanks again for your work!
